Chained.h

```diff
@@ -2,17 +2,206 @@
+            if(print) {
+                std::cout << i << ": |" << (*hashmap)[i].size() << "| ";
+            }
+            elements.push_back(static_cast<uint64_t>((*hashmap)[i].size()));
+            if((*hashmap)[i].empty()) {
+                ++numWith0;
+                collisions.push_back(0);
+            } else {
+                collisions.push_back(static_cast<uint64_t>((*hashmap)[i].size()-1));
+                if((*hashmap)[i].size() == 1){
+                    ++numWith1;
+                }
+                else{
+                    ++numWithMore;
+                }
+            }
+            for(auto kvp : (*hashmap)[i]){
+                if(print) {
+                    printKeyValuePair(kvp);
+                }
+            }
+            if(print) {
+                std::cout << "\n";
+            }
+        }
+        std::cout << std::endl;
+
+        //stats output
+        if(elements.empty()){
+            std::cout << "No Elements\n";
+        } else {
+            std::sort(elements.begin(), elements.end());
+            Statistic* elementStats = evaluateStats(&elements);
+            std::cout << "Total Elements: " << elementStats->total << "\n";
+            std::cout << "Mean Elements: " << elementStats->mean << "\n";
+            std::cout << "Stdev Elements: " << elementStats->stdev << "\n";
+
+            std::sort(collisions.begin(), collisions.end());
+            Statistic* collisionStats = evaluateStats(&collisions);
+            std::cout << "Total Collisions: " << collisionStats->total << "\n";
+            std::cout << "Mean Collisions: " << collisionStats->mean << "\n";
+            std::cout << "Stdev Collisions: " << collisionStats->stdev << "\n";
+
+            std::cout << "Total Number of Lists: " << hashmapSize << "\n";
+            std::cout << "Number of Lists With Zero Elements: " << numWith0 << "\n";
+            std::cout << "Number of Lists with One Element: " << numWith1 << "\n";
+            std::cout << "Number of Lists with more than One Element: " << numWithMore << "\n";
+            //I could make a display of all element sizes. Box plot? Scatterplot?
+
+            std::cout << "Number of elements in the biggest List: " << elements[elements.size()-1] << "\n";
+            std::cout << "Number of elements in the smallest List: " << elements[0] << "\n";
+        }
+    }
 };
+
+template<>
+uint64_t ChainedMap<std::string, uint64_t>::defaultGetReturn() {
+    return -1;
+}
+template<>
+Product* ChainedMap<uint64_t, Product*>::defaultGetReturn(){
+    return nullptr;
+}
+
+
+template<>
+uint64_t ChainedMap<std::string, uint64_t>::hash(std::string k){
+    return hasher(k)%hashmapSize;
+}
+template<>
+uint64_t ChainedMap<uint64_t, Product*>::hash(uint64_t k){
+    //this is random, inefficient math that exists to somewhat randomize placement
+    return (static_cast<uint64_t>(floor((sqrt(static_cast<double>(hasher(k)))*(k+3)))))%hashmapSize;
```

Comment:
Bad!
Design of hash functions is not something to be messed with.  You can choose a bad function for one choice of seed/input can fail catastrophically.

Using std::hash should be good enough.  You can also google hash functions.  There are decent ones that are short that you can inline.  We use xxHash, which is fast and good.

Reply:
The problem with std::hash and an integer is it just returns the integer [ f(i) = i ] so it isn't random at all. Is that okay?

Reply:
Well, that is good to investigate why you don't want it :)

We use xxHash at katana.  There are others.
https://github.com/Cyan4973/xxHash

MurmurHash is pretty simple.
https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp

Reply:
I think std::hash is okay, because it is the best unique hash for a given integer

Reply:
It depends on your input.  You should be able to generate a really bad input data set for that hash function, which is bad.